src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs

```diff
@@ -66,49 +53,29 @@ internal partial class Session
                             return currentModel;
                         }
 
-                        if (triggerInfo.TriggerReason == SignatureHelpTriggerReason.RetriggerCommand)
-                        {
-                            if (currentModel == null ||
-                                (triggerInfo.TriggerCharacter.HasValue && !currentModel.Provider.IsRetriggerCharacter(triggerInfo.TriggerCharacter.Value)))
-                            {
-                                return currentModel;
-                            }
-                        }
+                        var service = document.GetLanguageService<SignatureHelpService>();
 
-                        // first try to query the providers that can trigger on the specified character
-                        var result = await ComputeItemsAsync(matchedProviders, caretPosition, triggerInfo, document, cancellationToken).ConfigureAwait(false);
-                        var provider = result.Item1;
-                        var items = result.Item2;
-
-                        if (provider == null)
+                        var signatureList = await service.GetSignaturesAsync(document, caretPosition, trigger, cancellationToken: cancellationToken).ConfigureAwait(false);
+                        if (signatureList?.Provider == null)
                         {
-                            // no match, so now query the other providers
-                            result = await ComputeItemsAsync(unmatchedProviders, caretPosition, triggerInfo, document, cancellationToken).ConfigureAwait(false);
-                            provider = result.Item1;
-                            items = result.Item2;
-
-                            if (provider == null)
-                            {
-                                // the other providers didn't produce items either, so we don't produce a model
-                                return null;
-                            }
+                            return null;
                         }
 
                         if (currentModel != null &&
-                            currentModel.Provider == provider &&
-                            currentModel.GetCurrentSpanInSubjectBuffer(disconnectedBufferGraph.SubjectBufferSnapshot).Span.Start == items.ApplicableSpan.Start &&
-                            currentModel.ArgumentIndex == items.ArgumentIndex &&
-                            currentModel.ArgumentCount == items.ArgumentCount &&
-                            currentModel.ArgumentName == items.ArgumentName)
+                            currentModel.Provider == signatureList.Provider &&
```

Comment:
completion lists don't hold onto the provider do they?  We want these to be just data.  Should this just be some string property on the signature list?